cdap-cli/src/main/java/io/cdap/cdap/cli/command/StartProgramCommand.java

```diff
@@ -66,19 +71,49 @@ public void perform(Arguments arguments, PrintStream output) throws Exception {
       // run with stored runtime args
       programClient.start(programId, isDebug, null);
       runtimeArgsString = SPACE_EQUALS_JOINER.join(programClient.getRuntimeArgs(programId));
-      output.printf("Successfully started %s '%s' of application '%s.%s' with stored runtime arguments '%s'\n",
+      if (waitForRunning(output, programClient, programId)) {
+        output.printf("Successfully started %s '%s' of application '%s.%s' with stored runtime arguments '%s'\n",
                     elementType.getName(), programName, appName, appVersion, runtimeArgsString);
+      } else {
+        output.printf("Could not start %s '%s' of application '%s.%s' with stored runtime arguments '%s'\n",
+                elementType.getName(), programName, appName, appVersion, runtimeArgsString);
+      }
     } else {
       // run with user-provided runtime args
       Map<String, String> runtimeArgs = ArgumentParser.parseMap(runtimeArgsString,
                                                                 ArgumentName.RUNTIME_ARGS.toString());
       programClient.start(programId, isDebug, runtimeArgs);
-      output.printf("Successfully started %s '%s' of application '%s.%s' with provided runtime arguments '%s'\n",
+      if (waitForRunning(output, programClient, programId)) {
+        output.printf("Successfully started %s '%s' of application '%s.%s' with provided runtime arguments '%s'\n",
                     elementType.getName(), programName, appName, appVersion, runtimeArgsString);
+      } else {
+        output.printf("Could not start %s '%s' of application '%s.%s' with provided runtime arguments '%s'\n",
+                elementType.getName(), programName, appName, appVersion, runtimeArgsString);
+      }
     }
 
   }
 
+  private Boolean waitForRunning(PrintStream printStream, ProgramClient programClient, ProgramId programId)
+          throws InterruptedException, IOException, ProgramNotFoundException, UnauthenticatedException {
+    // 1sec waiting
+    int waitingMillis = 1000;
+    int count = 0;
+    while (count < Integer.MAX_VALUE) {
+      Thread.sleep(waitingMillis);
+      String status = programClient.getStatus(programId);
+      printStream.printf(" ..");
+      if (status.equals(ProgramStatus.RUNNING.toString())) {
```

Comment:
A loop check like this is subject to race condition. E.g. a program might be completed before this check, hence will run into a technically infinite loop (max int seconds). Also, there could also be concurrent run.

Reply:
Rightly pointed, this is prone to race condition.
Please suggest another fault proof approach. 

I can only think of :
a) include programstatus as stopped in check. (if it already ran before next iteration it will be in stopped state) 
b) fetching all run records and check status there by looping like this only. Will it work ?

Reply:
The proper way to fix it is to have the REST endpoint returns the runId of the program that get started. The run id is actually available in the backend service (https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java#L387), but the `ProgramLifecycleHttpHandler` currently just throwing it away.

Changing the REST API would requires API review to make sure it is backward compatible.

Reply:
Returning runid is a longstanding issue (https://issues.cask.co/browse/CDAP-7641), though it doesn't look like a big change.

Reply:
Additionally, ProgramClient already has a `waitForStatus` method, which we should leverage instead of writing this loop here.